kaspresso/src/main/kotlin/com/kaspersky/kaspresso/autoscroll/AutoScrollFallbackProviderImpl.kt

```diff
@@ -0,0 +1,129 @@
+import com.kaspersky.kaspresso.params.AutoScrollParams
+import org.hamcrest.Matcher
+import org.hamcrest.Matchers
+import androidx.test.espresso.matcher.ViewMatchers.isDisplayingAtLeast
+
+
+class AutoScrollFallbackProviderImpl(
+    private val params: AutoScrollParams,
+    private val logger: UiTestLogger
+) : AutoScrollProvider<ViewInteraction> {
+
+    /**
+     * Invokes the given [action] and calls [scroll] if it fails. Helps in cases when test fails because of the
+     * need to scroll to interacted view.
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     *
+     * @throws Throwable if the exception caught while invoking [action] is not allowed via [params].
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> withAutoScroll(interaction: ViewInteraction, action: () -> T): T {
+        return try {
+            action.invoke()
+        } catch (error: Throwable) {
+            if (error.isAllowed(params.allowedExceptions)) {
+                return scroll(interaction, action, error)
+            }
+            throw error
+        }
+    }
+
+    /**
+     * Performs the autoscrolling functionality. Performs scroll and re-invokes the given [action].
+     *
+     * @param interaction the instance of [ViewInteraction] interface to perform actions and assertions.
+     * @param action the actual action on the interacted view.
+     * @param cachedError the error to be thrown if autoscroll would not help.
+     *
+     * @throws cachedError if autoscroll action did not help.
+     * @return the result of [action] invocation.
+     */
+    @Throws(Throwable::class)
+    override fun <T> scroll(interaction: ViewInteraction, action: () -> T, cachedError: Throwable): T {
+        return try {
+            interaction.perform(FallbackScrollToAction())
+            logger.i("View autoScroll fallback successfully performed.")
+            action.invoke()
+        } catch (error: Throwable) {
+            throw cachedError
+        }
+    }
+}
+
+class FallbackScrollToAction : ViewAction {
+    override fun getConstraints(): Matcher<View> {
+        return Matchers.allOf(
+            ViewMatchers.withEffectiveVisibility(ViewMatchers.Visibility.VISIBLE),
+            ViewMatchers.isDescendantOfA(
+                Matchers.anyOf(
+                    ViewMatchers.isAssignableFrom(NestedScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(HorizontalScrollView::class.java),
+                    ViewMatchers.isAssignableFrom(ListView::class.java)
+                )
+            )
+        )
+    }
+
+    private tailrec fun findFirstParentScrollView(view: View?): View? {
+        return if (view is ScrollView || view is NestedScrollView || view is ListView) {
```

Comment:
Should add HorizontalScrollView... I also wonder if the padding issue also happens with RecyclerViews??
In that case this solution will not work on them, we will need more code to handle that situation